zebra-chain/src/work/difficulty.rs

```diff
@@ -320,3 +323,47 @@ impl AddAssign for Work {
         *self = *self + rhs;
     }
 }
+
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+/// Partial work used to track relative work in non-finalized chains
+pub struct PartialCumulativeWork(usize);
+
+impl std::ops::Add<Work> for PartialCumulativeWork {
+    type Output = PartialCumulativeWork;
+
+    fn add(self, rhs: Work) -> Self::Output {
+        let result = (self.0 as u128)
+            .checked_add(rhs.0)
+            .expect("Work values do not overflow")
+            .try_into()
+            .expect("......... uhhhh, suddenly wondering why work is a u128");
+
+        PartialCumulativeWork(result)
+    }
+}
+
+impl std::ops::AddAssign<Work> for PartialCumulativeWork {
+    fn add_assign(&mut self, rhs: Work) {
+        *self = *self + rhs;
+    }
+}
+
+impl std::ops::Sub<Work> for PartialCumulativeWork {
+    type Output = PartialCumulativeWork;
+
+    fn sub(self, rhs: Work) -> Self::Output {
+        let result = (self.0 as u128)
+            .checked_sub(rhs.0)
+            .expect("Work values do not overflow")
+            .try_into()
+            .expect("......... uhhhh, suddenly wondering why work is a u128");
```

Comment:
Subtractions can underflow - this is actually a very important check, because it makes sure we are only subtracting values we've previously added.

```suggestion
        let result = self.0
            .checked_sub(rhs.0)
            .expect("PartialCumulativeWork values do not underflow: all subtracted Work values must have been previously added to the PartialCumulativeWork")
```